plugins/net_plugin/net_plugin.cpp

```diff
@@ -4040,92 +3948,265 @@ namespace eosio {
+         auto cend = connections.end();
+         // do we remember the previous source?
+         if (current) {
+            //try to find it in the list
+            cptr = connections.find( current );
+            cend = cptr;
+            if( cptr == connections.end() ) {
+               // not there - must have been closed! cend is now connections.end, so just flatten the ring.
+               new_sync_source.reset();
+               cptr = connections.begin();
+            } else {
+               // was found - advance the start to the next. cend is the old source.
+               if( ++cptr == connections.end() && cend != connections.end() ) {
+                  cptr = connections.begin();
+               }
+            }
+         }
+
+         // scan the list of peers looking for another able to provide sync blocks.
+         if( cptr != connections.end() ) {
+            auto cstart_it = cptr;
+            do {
+            // select the first one which is current and has lib above current and break out.
+            if( !(*cptr)->is_transactions_only_connection() && (*cptr)->current() ) {
+               std::lock_guard<std::mutex> g_conn( (*cptr)->conn_mtx );
+               // TODO: change to a better heuristic than lib >= sync_known_lib_num
+               if( (*cptr)->last_handshake_recv.last_irreversible_block_num >= sync_known_lib_num ) {
+                  new_sync_source = *cptr;
+                  break;
+               }
+            }
+            if( ++cptr == connections.end() )
+               cptr = connections.begin();
+            } while( cptr != cstart_it );
+         }
+
+         // no need to check the result, either source advanced or the whole list was checked and the old source is reused.
+      }
+      return new_sync_source;
+   }
+
+   std::optional<connection_status> connections_manager::status( const string& host )const {
+      std::shared_lock g( connections_mtx );
+      auto con = find_connection_impl( host );
+      if( con ) {
          return con->get_status();
+      }
       return {};
    }
 
-   vector<connection_status> net_plugin::connections()const {
+   vector<connection_status> connections_manager::connection_statuses()const {
       vector<connection_status> result;
-      std::shared_lock<std::shared_mutex> g( my->connections_mtx );
-      result.reserve( my->connections.size() );
-      for( const auto& c : my->connections ) {
+      std::shared_lock g( connections_mtx );
+      result.reserve( connections.size() );
+      for( const auto& c : connections ) {
          result.push_back( c->get_status() );
       }
       return result;
    }
 
    // call with connections_mtx
-   connection_ptr net_plugin_impl::find_connection( const string& host )const {
-      for( const auto& c : connections )
-         if( c->peer_address() == host ) return c;
+   connection_ptr connections_manager::find_connection_impl( const string& host )const {
+      for( const auto& c : connections ) {
```

Comment:
This loop would be faster on a `flat_set`. 

I know this is not related to your change, but instead of having:

```
      std::set<connection_ptr>  connections;
      chain::flat_set<string>   supplied_peers;
```

why don't we have:

```
     std::unordered_map<string, std::optional<connection_ptr>> peers;
```

Reply:
Because there are incoming connections in `connections` that are not supplied via command-line/config.ini or `/v1/chain/connect`.